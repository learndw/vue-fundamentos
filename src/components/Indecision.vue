<template>
  <h1>Indesicion</h1>
  <img v-if="image" :src="image" />
  <br />
  <div class="bg-dark"></div>
  <input type="text" placeholder="Haz una pregunta" v-model="question" />
  <p>Recuerda terminar con un signo de interrogacion (?)</p>

  <div>
    <h2>{{ question }}</h2>
    <h1>{{ answer }}</h1>
  </div>
</template>

<script>
export default {
  name: "Indesicion",
  data() {
    return {
      question: null,
      answer: null,
      image: "https://via.placeholder.com/250",
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      this.answer = "Pensando ...";

      try {
        const { answer, image } = await fetch("https://yesno.wtf/api").then(
          (r) => r.json()
        );

        this.image = image;
        switch (answer) {
          case "yes":
            this.answer = "Si";
            break;
          case "no":
            this.answer = "No";
            break;
        }
      } catch (error) {
        console.log('Indesicion Component',error);
        this.answer = "Error al cargar";
        this.image = null;
      }
    },
  },
  //Observable que sta pendiente del valor actual y anterior de una propiedad
  watch: {
    question(value, oldValue) {
      this.isValidQuestion = false;
      console.log(value);
      if (!value.includes("?")) return;
      this.getAnswer();
    },
  },
};
</script>


