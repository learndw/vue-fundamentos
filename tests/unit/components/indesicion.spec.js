import { shallowMount } from "@vue/test-utils"
import Indecision from "@/components/Indecision"

describe('Indecision component snapshot', () => {
     let wrapper;
     let clgSpy;

     //Evitar error de fetch
     global.fetch = jest.fn(() => Promise.resolve({
          json: () => Promise.resolve({

               "answer": "yes",
               "forced": false,
               "image": "https://yesno.wtf/assets/yes/5-64c2804cc48057b94fd0b3eaf323d92c.gif"

          })
     }));

     beforeEach(() => {
          wrapper = shallowMount(Indecision);
          //Esta pendiente una accion en algun elemento del componente en este cao de cuantas veces se ejecuta el console log de watch
          clgSpy = jest.spyOn(console, 'log');

          jest.clearAllMocks();
     });

     test('Debe hacer match con el snapshot', () => {
          expect(wrapper.html()).toMatchSnapshot();
     })

     test('Escribir en el input no debe disparar nada (console.log)', async () => {

          //Obtener un metodo definido en methods en el componente de vue
          const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');

          //Obtiene el input
          const input = wrapper.find('input');

          //Asigna un valor
          await input.setValue('Algun valor');

          //Cuantas veces se espera que se ejecute en este caso el metodo watch la parte de console.log
          expect(clgSpy).toHaveBeenCalledTimes(1);

          //Se espera que el metodo no ha sido ejectuado por que el texto no conteien el signo de interrogacion
          expect(getAnswerSpy).not.toHaveBeenCalled();
     })

     test('Escribir en el input el simbolo de "?" para disparar el metodo getAnswer', async () => {

          //Obtener un metodo definido en methods en el componente de vue
          const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');

          //Obtiene el input
          const input = wrapper.find('input');

          //Asigna un valor
          await input.setValue('Va a llover?');

          //Se espera que el metodo no ha sido ejectuado por que el texto no conteien el signo de interrogacion
          expect(getAnswerSpy).toHaveBeenCalled();
     })

     test('Pruebas en getAnswer', async () => {
          await wrapper.vm.getAnswer();

          const img = wrapper.find('img');

          expect(img.exists()).toBeTruthy();

          expect(wrapper.vm.image).toBe('https://yesno.wtf/assets/yes/5-64c2804cc48057b94fd0b3eaf323d92c.gif');

          expect(wrapper.vm.answer).toBe('Si');
     })

     test('Pruebas en getAnswer - Fallo en API', async () => {

          fetch.mockImplementationOnce(() => Promise.reject('API is Down'));

          await wrapper.vm.getAnswer();

          const img = wrapper.find('img');

          expect(img.exists()).toBeFalsy();

          expect(wrapper.vm.answer).toBe('Error al cargar');
     })
})