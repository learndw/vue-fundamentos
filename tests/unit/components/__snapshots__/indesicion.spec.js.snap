// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component snapshot Debe hacer match con el snapshot 1`] = `
<h1>Indesicion</h1>
<img src="https://via.placeholder.com/250">
<br>
<div class="bg-dark"></div>
<input type="text" placeholder="Haz una pregunta">
<p>Recuerda terminar con un signo de interrogacion (?)</p>
<div>
  <h2></h2>
  <h1></h1>
</div>
`;
